Pyrado/pyrado/algorithms/meta/sprl.py

```diff
@@ -0,0 +1,573 @@
+                    distribution.distribution, target_distribution.distribution
+                )
+                grads = to.autograd.grad(kl_divergence, distribution.parameters())
+
+                return (
+                    kl_divergence.detach().numpy(),
+                    np.concatenate([g.detach().numpy() for g in grads]),
+                )
+
+            objective_fn = objective
+
+        # If we have never reached the performance threshold we optimize just based on the kl constraint
+        elif not self._performance_lower_bound_reached:
+            constraints = [kl_constraint]
+
+            # Now we optimize on the expected performance, meaning maximizing it
+            def objective(x):
+                distribution = previous_distribution.from_stacked(x)
+                performance = self._compute_expected_performance(distribution, contexts, contexts_old_log_prob, values)
+                grads = to.autograd.grad(performance, distribution.parameters())
+
+                return (
+                    -performance.detach().numpy(),
+                    -np.concatenate([g.detach().numpy() for g in grads]),
+                )
+
+            objective_fn = objective
+
+        if objective_fn:
+            result = minimize(
+                objective_fn,
+                x0,
+                method="trust-constr",
+                jac=True,
+                constraints=constraints,
+                options={"gtol": 1e-4, "xtol": 1e-6},
+                bounds=bounds,
+            )
+        if result and result.success:
+            self._adapt_parameters(result.x)
+
+        # We have a result but the optimization process was not a success
+        elif result:
+            old_f = objective_fn(previous_distribution.get_stacked())[0]
+            constraints_satisfied = all((const.lb <= const.fun(result.x) <= const.ub for const in constraints))
+
+            std_ok = bounds is None or (np.all(bounds.lb <= result.x)) and np.all(result.x <= bounds.ub)
+
+            if constraints_satisfied and std_ok and result.fun < old_f:
+                self._adapt_parameters(result.x)
+            else:
+                print(f"Update unsuccessful, keeping old values spl parameters")
+
+    def reset(self, seed: int = None):
+        # Forward to subroutine
+        self._subroutine.reset(seed)
+
+    def save_snapshot(self, meta_info: dict = None):
+        super().save_snapshot(meta_info)
+
+        if meta_info is None:
+            # This algorithm instance is not a subroutine of another algorithm
+            self._subroutine.save_snapshot(meta_info)
+
+    # This is the same operation as we do in part1 of the previous computation of the context loss
+    # But since we want to use it as a constraint, we have to calculate it here explicitly
+    def _compute_expected_performance(
+        self, distribution: MultivariateNormalWrapper, context: to.Tensor, old_log_prop: to.Tensor, values: to.Tensor
+    ) -> to.Tensor:
```

Comment:
Please add a one line doc string """..."""

Reply:
Done in 3e1479f683